test/distributed/test_distributed_spawn.py

```diff
@@ -0,0 +1,100 @@
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+import os
+import sys
+import unittest
+
+import torch.distributed as dist
+from torch._utils_internal import (
+    TEST_MASTER_ADDR as MASTER_ADDR,
+    TEST_MASTER_PORT as MASTER_PORT,
+)
+from torch.testing._internal.common_distributed import (
+    TEST_SKIPS,
+    MultiProcessTestCase,
+    initialize_temp_directories,
+    cleanup_temp_dir,
+)
+from torch.testing._internal.common_utils import run_tests, TEST_WITH_ASAN, NO_MULTIPROCESSING_SPAWN
+from torch.testing._internal.distributed.distributed_test import Barrier, _DistTestBase
+
+if not dist.is_available():
+    print("Distributed not available, skipping tests", file=sys.stderr)
+    sys.exit(0)
+
+BACKEND = os.environ["BACKEND"]
+INIT_METHOD = os.getenv("INIT_METHOD", "env://")
+
+
+if BACKEND == "gloo" or BACKEND == "nccl":
+    WORLD_SIZE = os.environ["WORLD_SIZE"]
+
+    @unittest.skipIf(
+        TEST_WITH_ASAN, "Skip ASAN as torch + multiprocessing spawn have known issues"
+    )
+    @unittest.skipIf(
+        NO_MULTIPROCESSING_SPAWN, "Spawn not available, skipping tests."
+    )
+    class TestDistBackendWithSpawn(MultiProcessTestCase, _DistTestBase):
+        @property
+        def init_method(self):
+            return "file://{file_name}".format(file_name=self.file_name)
+
+        # Needed since MultiProcessTestCase assumes a world_size of 4, but we
+        # run these tests under other various world_sizes.
+        @property
+        def world_size(self):
+            return os.environ["WORLD_SIZE"]
+
+        @classmethod
+        def setUpClass(cls):
+            os.environ["MASTER_ADDR"] = str(MASTER_ADDR)
+            os.environ["MASTER_PORT"] = str(MASTER_PORT)
+            os.environ["WORLD_SIZE"] = str(WORLD_SIZE)
+            super().setUpClass()
+
+        def setUp(self):
+            super().setUp()
+            # initialize Barrier.
+            initialize_temp_directories()
+            Barrier.init()
+            self._spawn_processes()
```

Comment:
Don't we only need to override the `setUp` method in this class? Shouldn't the other methods move to the base class and be shared between fork and spawn?

Reply:
Good point, it looks like we can dedupe most of this with the fork version and put the shared code in `_DistTestBase`.